onnxruntime/core/session/inference_session.h

```diff
@@ -682,7 +677,7 @@ class InferenceSession {
 
   // Immutable state for each op in the model. Shared by all executors.
   // It has a dependency on execution_providers_.
-  std::unique_ptr<SessionState> session_state_;
+  std::optional<SessionState> session_state_;
```

Comment:
An InferenceSession will _always_ contain SessionState. It's not optional. Making this optional conveys different semantics. It says you can have a session with or without a session state.

Reply:
It is a kind of "Delayed Initialization".

Reply:
std::unique_ptr and std::optional works very similarly, except std::optional reduces one malloc call.

Reply:
Alright. A class having an optional member tells me that it's optional for an instance of this class to have this member (which is obviously not the case for SessionState). However, I see that many articles are using it as 'hack' to save a heap allocation when initialization is delayed. Looks like it's already an established pattern :-(.

Reply:
Yes, it is a pattern. However, there is a couple of considerations here at play.
1) if the sizeof(T) is big, we do not want to blow the stack when the InferenceSession is instantiated on the stack. Using optional for SessionState means it contributes directly to the size of the session. Probably, not an issue here.
2) We care about heap allocations because it is considered expensive in C++ and makes latency variance wider, and that is bad for RT users. However, it also means that we care about allocations made mostly on the inference path. Things that take place during model initialization etc, are not a big concern for us, although I did some replacements as a part of refactoring.